smartapps/curb/curb-energy-monitor.src/curb-energy-monitor.groovy

```diff
@@ -0,0 +1,477 @@
+/**
+ *
+ *  Copyright 2017 Curb
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License. You may obtain a copy of the License at:
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+
+include 'asynchttp_v1'
+
+definition(
+    name: "CURB Energy Monitor",
+    namespace: "curb",
+    author: "Curb",
+    description: "Gain insight into energy usage throughout your home.",
+    category: "Green Living",
+    iconUrl: "http://energycurb.com/wp-content/uploads/2015/12/curb-web-logo.png",
+    iconX2Url: "http://energycurb.com/wp-content/uploads/2015/12/curb-web-logo.png",
+    iconX3Url: "http://energycurb.com/wp-content/uploads/2015/12/curb-web-logo.png",
+    singleInstance: true
+) {
+    appSetting "clientId"
+    appSetting "clientSecret"
+}
+
+preferences {
+    page(
+        name: "auth",
+        title: "Curb",
+        nextPage: "",
+        content: "authPage",
+        uninstall: true)
+}
+
+mappings {
+    path("/oauth/initialize") {
+        action: [GET: "oauthInitUrl"]
+    }
+    path("/oauth/callback") {
+        action: [GET: "callback"]
+    }
+}
+
+def installed() {
+    log.debug "Installed with settings: ${settings}"
+}
+
+def updated() {
+    log.debug "Updated with settings: ${settings}"
+    removeChildDevices(getChildDevices())
+    unsubscribe()
+    initialize()
+}
+
+def initialize() {
+    log.debug "Initializing"
+    unschedule()
+    refreshAuthToken()
+    updateSelectedLocationId()
+    getDevices()
+    runEvery1Minute(getAllData)
+}
+
+def uninstalled() {
+    log.debug "Uninstalling"
+    removeChildDevices(getChildDevices())
+}
+
+def authPage() {
+    if (!state.accessToken) {
+        state.accessToken = createAccessToken()
+    }
+
+    if (state.authToken) {
+        return dynamicPage(name: "auth", title: "Login Successful", nextPage: "", install: true, uninstall: true) {
+            section() {
+                paragraph("Select your CURB Location")
+                input(
+                    name: "curbLocation",
+                    type: "enum",
+                    title: "CURB Location",
+                    options: state.locationNames
```

Comment:
There is a race condition being built here if the callback takes longer to execute, particularly since it's two network levels deep (the callback and then another request to get locations). It would be easier to say `options: curbLocations` - this will call `getCurbLocations`. `getCurbLocations()` should return a hash map, with the key as the location id and the value as the location name. Then the user can select one from this map.